vendor/k8s.io/kubernetes/openshift-kube-apiserver/admission/scheduler/podtolerationsrestriction/admission.go

```diff
@@ -0,0 +1,156 @@
+
+func (p *podTolerations) Validate(attr admission.Attributes, _ admission.ObjectInterfaces) error {
+	if attr.GetResource().GroupResource() != coreapi.Resource("pods") && attr.GetKind().GroupKind() != coreapi.Kind("Pod") && len(attr.GetSubresource()) == 0 {
+		return nil
+	}
+	pod, ok := attr.GetObject().(*coreapi.Pod)
+	if !ok {
+		return admission.NewForbidden(attr, fmt.Errorf("unexpected Type %T", attr.GetObject()))
+	}
+	return p.validatePodTolerations(attr, pod.Spec)
+}
+
+func (p *podTolerations) validatePodTolerations(attr admission.Attributes, podSpec coreapi.PodSpec) error {
+	// if a pod has no tolerations, then no need to validate it's tolerations
+	if len(podSpec.Tolerations) == 0 {
+		return nil
+	}
+	tolerations := podSpec.Tolerations
+	if len(podSpec.ServiceAccountName) == 0 {
+		return admission.NewForbidden(attr, fmt.Errorf("service account is not provided. The restrict pod tolerations plugin is completely lost"))
+	}
+	for _, toleration := range tolerations {
+		allow := p.IsAuthorized(podSpec.ServiceAccountName, attr, toleration)
+		if !allow {
+			return admission.NewForbidden(attr, fmt.Errorf("requested user %v doesn't have permission to provide the given tolerations", podSpec.ServiceAccountName))
+		}
+	}
+	return nil
+}
+
+func (p *podTolerations) IsAuthorized(podSAName string, attr admission.Attributes, toleration coreapi.Toleration) bool{
+	var saInfo user.Info
+	saInfo = serviceaccount.UserInfo(attr.GetNamespace(), podSAName, "")
+	// check for hard-coded serviceaccount groups
+	SAInNamespace := serviceaccount.UserInfo(attr.GetNamespace(), "system:serviceaccounts:"+attr.GetNamespace(), "")
+	SAAcrossAllNamespace := serviceaccount.UserInfo("", "system:serviceaccounts", "")
+	return p.checkPodTolerationAccess(saInfo, attr, toleration) || p.checkPodTolerationAccess(SAInNamespace, attr, toleration) || p.checkPodTolerationAccess(SAAcrossAllNamespace, attr, toleration)
+}
+
+// buildTolerationsAuthorizationAttributes builds the tolerations attributes from the user
+func buildTolerationsAuthorizationAttributes(info user.Info, namespace string, toleration coreapi.Toleration) authorizer.Attributes {
+	verb := toleration.Operator
+	tolerationKey := toleration.Key
+	if len(toleration.Key) == 0 {
+		tolerationKey = masterToleration
+	}
+	authzAttr := authorizer.AttributesRecord{
+		User:            info,
+		Verb:            string(verb),
+		Namespace:       namespace,
+		Resource:        "tolerations",
+		APIGroup:        "toleration.scheduling.openshift.io",
+		ResourceRequest: true,
+		Name:            tolerationKey,
+	}
+	return authzAttr
+}
+
+
+// checkPodTolerationAccess checks if pods are allowed to have a particular toleration. In this case, we're focusing
+// on master tolerations. For example,
+/*
+tolerations:
+- effect: NoSchedule
+  key: node-role.kubernetes.io/master
+  operator: Exists
+
+are converted to
+
+apiGroup: toleration.scheduling.openshift.io
+Resource: toleration
+resourceName: <empty_string>/NoSchedule
```

Comment:
I don't think you understand why the `value/effect` in the resource name makes sense.  Also, you can't split on slash here.  For splitting, you probably want the effect first, right?

Empty names match all names.  Specified names match their specification.  This construct makes it easy to allow someone to tolerate a particular key with any value or any effect.  It makes it possible to enumerate the key/effect tuple as a unit.

Reply:
The above example is wrong. Please look at https://github.com/openshift/origin/pull/23427/files#diff-b49487e41f10910366174f767ef67f90R99-R137

Reply:
Ok, so, if someone wants to have toleration against any key with any value or any effect, they can use the https://github.com/openshift/origin/pull/23427/files#diff-b49487e41f10910366174f767ef67f90R121-R123